### PR TITLE
Fixes for building with picolibc

### PR DESCRIPTION
**Pull request by keith-packard**
I found a couple of minor issues in Zephyr's picolibc support:

 1. scrambled printfcb wrapper macros
 2. Need to use -fno-printf-return-value when using smaller printf variants to avoid gcc optimizations around sprintf return values.
 3. Zephyr wants errno values covered under the `__LINUX_ERRNO_EXTENSIONS__` definition.
 4. qemu_x86_tiny needs to handle picolibc like newlib and let it use any available heap for malloc.

### CODE REVIEW

File: cmake/compiler/gcc/compiler_flags.cmake [hunk 1]
```diff
@@ -111,6 +111,12 @@ if (NOT CONFIG_NEWLIB_LIBC AND
   set_compiler_property(APPEND PROPERTY nostdinc_include ${NOSTDINC})
 endif()
 
+if (CONFIG_PICOLIBC AND
+    NOT CONFIG_PICOLIBC_IO_FLOAT)
+  set_compiler_property(PROPERTY no_printf_return_value -fno-printf-return-value)
```
Comment (stephanosio) [inline cmake/compiler/gcc/compiler_flags.cmake:116]: there needs to be an entry for `no_printf_return_value` in `cmake/compiler/compiler_flags_template.cmake`.

**Review by stephanosio — CHANGES_REQUESTED** [review]

File: cmake/compiler/gcc/compiler_flags.cmake (hunk 1 above)
Comment (keith-packard) [reply cmake/compiler/gcc/compiler_flags.cmake:116]: thanks, added.

**Review by cfriedt — DISMISSED** [review]

File: cmake/compiler/gcc/compiler_flags.cmake [hunk 2]
```diff
@@ -111,6 +111,11 @@ if (NOT CONFIG_NEWLIB_LIBC AND
   set_compiler_property(APPEND PROPERTY nostdinc_include ${NOSTDINC})
 endif()
 
+if (CONFIG_PICOLIBC AND
+    NOT CONFIG_PICOLIBC_IO_FLOAT)
```
Comment (tejlmand) [inline cmake/compiler/gcc/compiler_flags.cmake:115]: setting of a compiler flag should not be dependent on a kconfig setting.
This is a valid flag for gcc, regardless of picolibc.

The decission whether the flag should be used during compilation should generally be taken based on Kconfig, but that is to be done elsewhere, like you already do here:
https://github.com/zephyrproject-rtos/zephyr/blob/0459bb2aca6be054c00bab07a5def195ab4d232e/CMakeLists.txt#L272-L279

Of course the current `if(<condition>)` might need to be adjusted according to the needs.

**Review by tejlmand — CHANGES_REQUESTED** [review]

**Review by cfriedt — APPROVED** [review]

**Review by stephanosio — APPROVED** [review]

**Review by dcpleung — APPROVED** [review]

**Review by tejlmand — APPROVED** [review]
lgtm

**Pull request merged**